cityofaustin/atd-knack-311
# Hello world: new Knack > 311 integration script

## Associated issue

- https://github.com/cityofaustin/atd-data-tech/issues/11175

This PR serves as a place to review this new script, which replaces these two legacy scripts:
- https://github.com/cityofaustin/atd-data-publishing/blob/production/transportation-data-publishing/data_tracker/esb_xml_gen.py
- https://github.com/cityofaustin/atd-data-publishing/blob/production/transportation-data-publishing/data_tracker/esb_xml_send.py

There is a related PR in the atd-prefect for deployment - [#53](https://github.com/cityofaustin/atd-prefect/pull/53).


## Steps to test

1. Prepare test records in the Data Tracker's test environment - `TEST - ESB ENV - AMD Data Tracker | 19 APR 2022`: navigate to the [`tmc_activities` object](https://builder.knack.com/atd/austin-transportation-data-tracker-esb-test-environment-20/records/objects/object_75). 

2. Select activities with ID's 1 & 2, and use the **Update** tool to set the **PROCESS_STATE_ESB** field to **READY_TO_SEND**

<img width="980" alt="Screen Shot 2023-02-03 at 11 56 02 AM" src="https://user-images.githubusercontent.com/14793120/216661783-3688f514-b5de-475c-8d02-a58e8c34551b.png">

3. Now you're ready to run the script. Pull (or build) the docker image: `atddocker/atd-knack-311:staging`

4. Configure an environment file which has the Data Tracker test environment credentials and the ESB - Test endpoint. See the "Environment for atd-knack-311" item in our password store.

5. Save a copy of self-signed SSL key and certificate in the root of this repository as `esb.pem` and `esb.cert`. You can locate these files by searching the password store for "atd-knack-311 self-signed". 

6. Run the script, mounting your local clone of the repo (with certs) into the container

```
$ docker run -it --rm --env-file env_file -v "$(pwd)":/app atddocker/atd-knack-311:staging python send_knack_messages_to_esb.py data-tracker
```

7. Refresh the Knack builder and observe that the two records you recently modified now have an **ESB_PROCESS_STATUS**  of **SENT**.
